imp(staking): replace bech32 address with evm hex address for staking precompile contract

luchenqun opened the pull request:
## Description
This PR will convert all bech32 addresses in precompiled contract staking to evm hex address. The implementation itself is not difficult, but too many unit tests need to be fixed. I originally wanted to fix one function such as delegate with one PR, but I found that there is a strong correlation between all functions, and it would be more efficient to fix them all at once. It is roughly planned to complete all tasks by the end of January next year.

Please help me review whether the changes to the StakingI.sol file are reasonable, because subsequent updates need to be based on the StakingI.sol file. @fedekunze @Vvaradinov 

More details see before https://github.com/evmos/evmos/pull/2122

0xstepit commented on precompiles/staking/integration_test.go:
```diff
@@ -65,6 +66,10 @@ var _ = Describe("Calling staking precompile directly", func() {
 		valAddr = s.validators[0].GetOperator()
 		valAddr2 = s.validators[1].GetOperator()
 
+		validatorAddr, err := sdk.ValAddressFromBech32(s.validators[0].GetOperator().String())
```
This will always error

0xstepit commented on precompiles/staking/integration_test.go:
```diff
@@ -65,6 +66,10 @@ var _ = Describe("Calling staking precompile directly", func() {
 		valAddr = s.validators[0].GetOperator()
 		valAddr2 = s.validators[1].GetOperator()
 
+		validatorAddr, err := sdk.ValAddressFromBech32(s.validators[0].GetOperator().String())
```
```suggestion
		validatorAddr, err := sdk.ValAddressFromBech32(valAddr.String())
```

luchenqun replied:
The above does not need to be reviewed for the time being. Review it together later when everything is completed. Just help me review whether the changes to **precompiles/staking/StakingI.sol** file are reasonable.

0xstepit commented:
Looks good @luchenqun, thanks for your contribution! I do agree it is a good idea to work with the evm hex in the precompile.

luchenqun commented:
> @GAtom22
> 
> Hello Tom, because I have changed the `string validator` to the `address validator`, one of the staking integration test **should execute queries for calltype** has failed. I think the solidity code here needs to be updated https://github.com/luchenqun/evmos/blob/224293dc3d6629a5640e2356e8abc423b3919142/precompiles/staking/testdata/StakingCaller.sol#L352-L393 But I am not particularly familiar with this part of the code , can you help me update this part of the code?

I modified it based on my intuition, and it seems to be OK. LOL

luchenqun commented:
@fedekunze @Vvaradinov It took an hour to implement the feature, but a day to fix the unit tests. Hahaha. Now you can start review the updated code.

hanchon commented on precompiles/staking/testdata/StakingCaller.sol:
```diff
@@ -334,15 +332,12 @@ contract StakingCaller {
         } else if (calltypeHash == keccak256(abi.encodePacked("callcode"))) {
             //Function signature
-            bytes4 sig = bytes4(keccak256(bytes("delegation(address,string)")));
+            bytes4 sig = bytes4(keccak256(bytes("delegation(address,address)")));
```
We need @GAtom22 to verify this change, I don't enough of solidity to be sure that it's the correct change (it looks good)

hanchon commented on precompiles/staking/staking_test.go:
```diff
@@ -96,27 +96,27 @@ func (s *PrecompileTestSuite) TestRequiredGas() {
 					staking.UndelegateMethod,
 					s.address,
-					s.validators[0].GetOperator().String(),
+					common.BytesToAddress(s.validators[0].GetOperator().Bytes()),
```
We need to a typescript library to convert from validator bech32 to hex in order to update the frontend. Without the lib the frontend will not be able to create staking transactions after upgrading the network with the new code.

hanchon commented:
Providing a Typescript lib to convert the validators addresses is a blocker to merge this PR. After having the lib ready we can finish the PR review and merge it

VictorTrustyDev commented:
Not Stale
